File: OpenBid/Mopub/ObjectiveC/SampleApp/SampleApp/EventHandlers/MoPub/Banner/POBBannerCustomEvent.m


#import "POBBannerCustomEvent.h"
#import <POBBid.h>
#import <POBRenderer.h>

@interface POBBannerCustomEvent()<POBAdRendererDelegate>
@property (nonatomic, strong) id<POBBannerRendering> bannerRenderer;
@end

@implementation POBBannerCustomEvent

- (void)requestAdWithSize:(CGSize)size customEventInfo:(NSDictionary *)info{
    if (self.localExtras == nil) {
        [self handleCustomEventFailure];
        return;
    }
    POBBid *bid = self.localExtras[@"pob_bid"];
    _bannerRenderer = [POBRenderer bannerRenderer];
    [_bannerRenderer setDelegate:self];
    [_bannerRenderer renderAdDescriptor:bid];
}

-(void)handleCustomEventFailure{
    NSError *error = [NSError errorWithDomain:kPOBErrorDomain code:POBErrorNoAds userInfo:@{NSLocalizedDescriptionKey:NSLocalizedString(@"Bid not available for custom event", comment: @"Bid not available for custom event")}];
    [self.delegate bannerCustomEvent:self didFailToLoadAdWithError:error];
}

#pragma mark - PMAdRendererDelegate
- (void)rendererDidRenderAd:(id)renderedAd forDescriptor:(id<POBAdDescriptor>)ad {
    UIView *rendererView = renderedAd;
    POBBid *bid = self.localExtras[@"pob_bid"];
    rendererView.frame = CGRectMake(0, 0, bid.size.width, bid.size.height);
    [self.delegate bannerCustomEvent:self didLoadAd:renderedAd];
}

- (void)rendererDidFailToRenderAdWithError:(NSError *)error forDescriptor:(id<POBAdDescriptor>)ad {
    [self.delegate bannerCustomEvent:self didFailToLoadAdWithError:error];
}

- (void)rendererWillPresentModal {
    [self.delegate bannerCustomEventWillBeginAction:self];
}

- (void)rendererDidDismissModal {
    [self.delegate bannerCustomEventDidFinishAction:self];
}

- (void)rendererWillLeaveApp {
    [self.delegate bannerCustomEventWillLeaveApplication:self];
}

- (UIViewController *)viewControllerForPresentingModal {
    return [self.delegate viewControllerForPresentingModalView];
}

- (void)dealloc {
    _bannerRenderer = nil;
}

@end

File: OpenBid/Mopub/Swift/SampleApp/SampleApp/EventHandlers/MoPub/Banner/POBBannerCustomEvent.swift

import UIKit

class POBBannerCustomEvent: MPBannerCustomEvent, POBAdRendererDelegate {
    
    var bannerRenderer: POBBannerRendering?
    
    override func requestAd(with size: CGSize, customEventInfo info: [AnyHashable : Any]!) {
        if localExtras == nil {
            self.handleCustomEventFailure()
            return
        }
        if let bid = localExtras["pob_bid"] as? POBBid {
            bannerRenderer = POBRenderer.bannerRenderer()
            bannerRenderer?.setDelegate(self)
            bannerRenderer?.renderAdDescriptor(bid)
        }else{
            self.handleCustomEventFailure()
        }
    }
    
    func handleCustomEventFailure() -> Void {
        self.delegate.bannerCustomEvent(self, didFailToLoadAdWithError: NSError(domain: kPOBErrorDomain, code: POBErrorCode.errorInternalError.rawValue, userInfo: [NSLocalizedDescriptionKey : NSLocalizedString("Bid not available for custom event", comment: "Bid not available for custom event")]))
    }
    
    // MARK: - POBAdRendererDelegate
    func rendererDidRenderAd(_ renderedAd: Any!, for ad: POBAdDescriptor!) {
        if let bid = localExtras["pob_bid"] as? POBBid {
            let rendererView = renderedAd as! UIView
            rendererView.frame = CGRect(x: 0, y: 0, width: (bid.size.width), height: (bid.size.height))
            self.delegate.bannerCustomEvent(self, didLoadAd: rendererView)
        }
    }
    
    func rendererDidFailToRenderAdWithError(_ error: Error!, for ad: POBAdDescriptor!) {
        self.delegate.bannerCustomEvent(self, didFailToLoadAdWithError: error)
    }
    
    func rendererWillLeaveApp() {
        self.delegate.bannerCustomEventWillLeaveApplication(self)
    }
    
    func rendererWillPresentModal() {
        self.delegate.bannerCustomEventWillBeginAction(self)
    }
    
    func rendererDidDismissModal() {
        self.delegate.bannerCustomEventDidFinishAction(self)
    }
    
    func viewControllerForPresentingModal() -> UIViewController! {
        return self.delegate.viewControllerForPresentingModalView()
    }
    
    deinit {
        bannerRenderer = nil
    }
}
